subsys/shell/shell_string_conv.c

```diff
@@ -0,0 +1,159 @@
+#include <ctype.h>
+#include <string.h>
+#include <stdlib.h>
+
+int shell_str2long(const struct shell *shl, char *str, int *val)
+{
+	int str_len = strlen(str);
+	int temp_val;
+	int idx = 0;
+
+	if (!strncmp(str, "-", 1)) {
+		idx = 1;
+	}
+
+	for (int i = idx; i < str_len; i++) {
+		if (isdigit((int)str[i])) {
+			continue;
+		}
+
+		shell_error(shl, "\"%s\" is not recognized as a valid signed decimal input", str);
+		return -EINVAL;
+	}
+
+	temp_val = strtol(str, NULL, 10);
+
+	if ((temp_val <= LONG_MIN) || (temp_val >= LONG_MAX)) {
+		shell_error(shl, "Passed input value \"%s\" is too small/large", str);
+		return -EINVAL;
+	}
+
+	*val = temp_val;
+	return 0;
+}
+
+int shell_str2ulong(const struct shell *shl, char *str, uint64_t *val)
+{
+	int str_len = strlen(str);
+	uint64_t temp_val;
+
+	for (int i = 0; i < str_len; i++) {
+		if (isdigit((int)str[i])) {
+			continue;
+		}
+
+		shell_error(shl, "\"%s\" is not recognized as a valid unsigned decimal input", str);
+		return -EINVAL;
+	}
+
+	temp_val = strtoul(str, NULL, 10);
+
+	if (temp_val >= ULONG_MAX) {
+		shell_error(shl, "Passed input value \"%s\" is too large", str);
+		return -EINVAL;
+	}
+
+	*val = temp_val;
+	return 0;
+}
+
+int shell_str2dbl(const struct shell *shl, const char *str, double *val)
+{
+	char deci_buf[12], frac_buf[11] = { 0 };
+	int decimal;
+	uint64_t frac;
+	double frac_dbl;
+	int idx = strcspn(str, ",.");
+
+	if (idx > (sizeof(deci_buf) - 1)) {
+		shell_error(shl, "Passed input value \"%s\" is too small/large", str);
+		return -EINVAL;
+	}
+
+	strncpy(deci_buf, str, idx);
+	strncpy(frac_buf, (str + (idx + 1)), MIN(sizeof(frac_buf) - 1, strlen(str + (idx + 1))));
+
+	if (shell_str2long(shl, deci_buf, &decimal)) {
+		return -EINVAL;
+	}
+
+	if (shell_str2ulong(shl, frac_buf, &frac)) {
+		return -EINVAL;
+	}
+
+	frac_dbl = (double)frac;
+
+	for (int i = 0; i < strlen(frac_buf); i++) {
+		frac_dbl /= 10;
+	}
+
+	*val = (str[0] == '-') ? ((double)decimal - frac_dbl) : ((double)decimal + frac_dbl);
+	return 0;
+}
+
+static bool hex_str_check(char *str, uint8_t str_len)
+{
+	if (str_len % 2) {
+		return false;
+	}
+
+	for (int i = 0; i < str_len; i++) {
+		if (!isxdigit((int)str[i])) {
+			return false;
+		}
+	}
+	return true;
+}
+
+uint8_t shell_hexstr2num(const struct shell *shl, char *str, uint8_t *buf, uint8_t buf_len)
+{
+	if (!buf || !buf_len) {
+		shell_error(shl, "Invalid buffer ref/size");
+		return 0;
+	}
+
+	char str_num[3] = { 0 };
+	int str_len = strlen(str);
+
+	if (!hex_str_check(str, str_len)) {
+		shell_error(shl, "Invalid hex string format");
+		return 0;
+	}
+
+	if ((str_len / 2) > buf_len) {
+		shell_error(shl, "Hex value is too large");
+		return 0;
+	}
+
+	for (int i = 0; i < str_len / 2; i++) {
+		strncpy(str_num, str + (i * 2), 2);
+		buf[i] = strtol(str_num, NULL, 16);
+	}
+
+	return str_len / 2;
+}
+
+int shell_str2bool(const struct shell *shl, char *str, bool *val)
+{
+	if (!strcmp(str, "on") || !strcmp(str, "enable")) {
```

Comment:
May you please let me know what can be a use case for this function?
I ask out of curiosity because usually options like `on` and `enable` are implemented as subcommands.

Reply:
It is just something that is inherited from the mesh shell module. These functions will eventually replace the string conversions inside the mesh shell module. On second thought it would be cleaner to leave this out and create a wrapper for it where it is needed.